### PR TITLE
copier: remove the legacy unbuffered copier

### PR DESCRIPTION
**Pull request by morgo**
## Why

The buffered DBLog-style copier has been the default since v0.15.0 (#908), and `move`/`sync` have always required it. A search across Block automation and downstream consumers (tern, strata) found nothing setting `--unbuffered`, so this retires the legacy `INSERT IGNORE .. SELECT` implementation rather than keep maintaining two copy paths. Net **−832 lines**.

## What changed

**Code**
- Delete `pkg/copier/unbuffered.go` and the `--unbuffered` flag.
- `CopierConfig` loses `Unbuffered` and the vestigial `TargetChunkTime` (chunk sizing lives entirely in the chunker); `NewCopier` drops its unused `*sql.DB` parameter and now requires a non-nil `Applier`.
- The migration runner's copy chunkers always size by `TargetChunkBytes` now; `--target-chunk-time` remains for the checksum. Also fixes the startup log printing the `TargetChunkTime` value under the `target-chunk-size` key.
- The autoscaling-vs-unbuffered downgrade warning is gone (the combination can no longer exist).
- **`Runner.Close()` now stops the applier.** `Run`'s teardown already stopped it (Stop is idempotent), but early-failure paths — and tests that step the copy incrementally — would otherwise leak the applier's worker goroutines.

**New API: `copier.ChunkCopier`**

The unbuffered copier's synchronous `CopyChunk` was the vehicle for every deterministic-stepping test (checkpoint/watermark ordering, binlog interleaving). The buffered copier now implements the same contract behind a small interface:

```go
type ChunkCopier interface {
    CopyChunk(ctx context.Context, chunk *table.Chunk) error
}
```

It auto-starts the applier (idempotent), applies the chunk, and waits on the applier callback — feedback reaches the chunker before it returns, so tests can complete chunks in a controlled order (e.g. 2, 1, 3) exactly as before.

**Displaced test coverage**
- Range-optimizer refusal (`range_optimizer_max_mem_size` → fatal) moved to `pkg/dbconn`, where the warning-code classification actually lives.
- The NULL→NOT NULL "unsafe warning" assertion is ported to the buffered path.
- Lossy-conversion behavior stays covered by `TestBufferedCopierDataTypeConversionError`.
- `TestResumeFromCheckpointPhantom` is **deleted, not ported**: per its own doc comment the phantom row only arises on the `INSERT IGNORE ... SELECT` recopy path; the buffered copier applies row images via REPLACE and has no equivalent scenario.
- All buffered/unbuffered test matrices collapse to single runs (rename-column, datatype, cutover-atomicity incl. the semisync build-tag variant, migration_test pairs).

**Docs**
- `docs/migrate.md` drops the `### unbuffered` section; README/AGENTS/package READMEs (copier, applier, table, throttler, migration) now describe the single-copier world, keeping short history notes where the contrast is still instructive.

## Verification

- `go build ./...`, `go vet` (including `semisync` and `singleversion` build tags), `gofmt`, `golangci-lint` (0 issues, uncapped).
- Against local MySQL 8.0.43: full `pkg/copier` (goleak-guarded), `pkg/change`, `pkg/move`, `pkg/datasync`; `pkg/dbconn -run TestRangeOptimizerRefusal`; focused `pkg/migration` runs covering the ported step-through tests (`TestCheckpoint`, `TestCheckpointRestore`, all 7 `TestE2EBinlogSubscribing*`), resume E2E, and every collapsed matrix (16 rename tests, enum/set reorder+drop, cutover atomicity, generated-columns/binary-checksum/charset pairs). All pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/throttler/README.md [hunk 1]
```diff
@@ -134,10 +134,10 @@ Both throttlers require an `IsAurora()` probe — which confirms `performance_sc
...
 
 ```go
-copier := &copier.Unbuffered{
-    Throttler: throttler,
-    // ... other config
-}
+config := copier.NewCopierDefaultConfig()
+config.Throttler = throttler
+// ... other config (Applier is required)
+c, err := copier.NewCopier(chunker, config)
```
[the remark this answers is not in the document]
Comment (morgo) [reply pkg/throttler/README.md:140]: 🤖 Fixed in 48fb5fa6 — the example now shows `config.Applier = rowApplier // required (non-nil); see pkg/applier`.

File: pkg/copier/buffered_test.go [hunk 2]
```diff
@@ -151,15 +152,15 @@ func TestBufferedCopierDataTypeConversionError(t *testing.T) {
 	require.NoError(t, t2.SetInfo(t.Context()))
 
 	cfg := NewCopierDefaultConfig()
-	cfg.TargetChunkTime = 10 // Small chunk time to create more chunks
 	cfg.Applier, err = applier.NewSingleTargetApplier(applier.Target{DB: db}, applier.NewApplierDefaultConfig())
 	require.NoError(t, err)
-	chunker, err := table.NewChunker(t1, table.ChunkerConfig{NewTable: t2, TargetChunkTime: cfg.TargetChunkTime, Logger: cfg.Logger})
+	// Tiny chunk-time target so the chunker creates many small chunks.
+	chunker, err := table.NewChunker(t1, table.ChunkerConfig{NewTable: t2, TargetChunkTime: 10, Logger: cfg.Logger})
```
[the remark this answers is not in the document]
Comment (morgo) [reply pkg/copier/buffered_test.go]: 🤖 Fixed in 48fb5fa6 — now `10 * time.Nanosecond` with the comment spelling out that the absurdly small target is intentional (it forces the chunker to create many small chunks).

File: pkg/copier/buffered.go [hunk 3]
```diff
@@ -63,8 +61,62 @@ type buffered struct {
...
+		}
+	}
+	if err := c.applier.Apply(ctx, chunk, rows, callback); err != nil {
+		return fmt.Errorf("failed to apply rows: %w", err)
+	}
+	// The applier guarantees the callback is invoked exactly once per Apply
+	// that returned nil (including on error or cancellation), so this cannot
+	// block forever.
+	<-done
+	return applyErr
```
[the remark this answers is not in the document]
Comment (morgo) [reply pkg/copier/buffered.go]: 🤖 Fixed in 48fb5fa6 — the wait is now a `select` on `done` / `ctx.Done()`. For the record, the in-tree appliers cannot strand this wait: `Apply` invokes the callback itself with `ctx.Err()` if cancelled mid-enqueue, workers convert cancellation into error completions, and the feedback coordinator deliberately drains until the completions channel closes (only `Stop` closes it, after workers exit) rather than exiting on `ctx.Done()`. The select is defense-in-depth against a future non-conforming applier; the comment notes that if the ctx branch fires, the callback (chunker feedback + metrics) may still run later on the coordinator goroutine.

**Review by aparajon — APPROVED** [review]
🤖 Approving. Retiring the second copy path is clearly right, and the part that could have gone wrong — losing the deterministic-stepping vehicle that the checkpoint and binlog-watermark tests were built on — is handled by giving the buffered copier the same synchronous contract rather than by loosening those tests. I checked the two production behaviour changes that hide inside the deletion and both hold up. Two findings, neither blocking.

**1 — the new `CopyChunk` contract is asserted only in another package, and only incidentally.** `ChunkCopier` exists for exactly one guarantee: chunker feedback has been sent by the time it returns. Every caller of it lives in `pkg/migration`, and every assertion on it is about something else — watermark ordering, checkpoint content, binlog interleaving. `pkg/copier` has no test of its own for the API it just introduced. That's the wrong failure mode to sign up for: if `Apply`'s callback semantics drift — a completion delivered on a different goroutine, an error path that skips the callback, an empty chunk short-circuiting — the symptom is a watermark test failing in `pkg/migration` with no obvious connection to the applier change that caused it. A focused test next to the implementation (copy one chunk, assert the chunker's watermark advanced before the call returned; copy an empty chunk; make the applier return an error and assert no feedback was sent) makes the contract fail where it's owned. Related, and worth a line in the doc comment rather than a code change: if the `ctx.Done()` branch ever does fire, the callback still runs later and feeds the chunker for a chunk whose caller was told it failed. The comment explains why that branch is unreachable today; it doesn't say what happens if it isn't.

**2 — none of the three PRs' CI runs says anything about the combination, and this one's interaction with #1138 is semantic, not textual.** All three conflict pairwise against each other (I trial-merged them: #1138↔#1140 collides in four files, #1139 collides with both). That's expected for parallel work, but the #1138 case is the one that won't be caught by resolving conflict markers: #1138 converts tests that this PR deletes outright (`TestResumeFromCheckpointPhantom`) and preserves `Unbuffered: true` scenarios whose struct field this PR removes, so a textually clean resolution can still fail to compile. Merging **#1140 first** is the cheaper order — it deletes the phantom test and the field, and #1138's rebase then shrinks by exactly the hunks that would otherwise need re-deriving. The reverse order leaves this PR re-deriving its deletions against converted call sites.

Action items, both follow-ups:

1. Add a `pkg/copier` test pinning `CopyChunk`'s feedback-before-return contract (plus the empty-chunk and apply-error paths).
2. Land this before #1138 and note the ordering on both, since neither PR's green CI covers the pair.

<details>
<summary>Verified</summary>

The two production behaviour changes bundled with the deletion both check out: `Runner.Close()` now appends `applier.Stop()`'s error, and `SingleTargetApplier.Stop` returns `nil` early when already stopped or never started, so the double-stop after a normal run cannot turn a successful `Close()` into an error — and on the shutdown path the applier's context is already cancelled, so the added wait doesn't lengthen teardown; the copy chunker "always sizes by `TargetChunkBytes`" is not a change for anyone real, since that branch was already taken whenever `Unbuffered` was false, which has been the default since v0.15.0; the startup-log fix is a genuine bug fix riding along (`target-chunk-size` was printing `TargetChunkTime`); `getCopyStats` derives progress from the chunker rather than a copier-local counter, so stepping via `CopyChunk` still advances it (only `rowsPerSecond`/ETA, which are `Run`-scoped, stay unset); the `applyErr` handoff is properly synchronized — the write precedes `close(done)` and the read follows `<-done`; and `NewCopier`'s new non-nil `Applier` requirement is an error rather than a nil-deref, with the `*sql.DB` parameter genuinely unused now that reads go through `chunk.Table.DB()`.

</details>

*This review was generated by Claude Code (claude-opus-5).*

**Comment by morgo** [comment]
🤖 (Claude, on Morgan's behalf) Both action items addressed:

**1 — contract test + doc line:** cb6f05e2 adds `pkg/copier/chunk_copier_test.go`: a `feedbackRecorder` wrapping the chunker pins feedback-delivered-before-return at the moment `CopyChunk` returns — for the row-carrying chunk (rows in the target, low watermark immediately queryable, no polling) and for the empty first chunk — and the apply-error path (target table dropped) asserts the error surfaces synchronously with no feedback sent for the failed chunk. The `CopyChunk` doc comment now also states the `ctx.Done()` consequence: if that branch ever fires, the callback still runs later on the coordinator goroutine and feeds the chunker for a chunk whose caller was told it failed.

**2 — ordering:** #1138 reached main first, so this PR took the expensive side of the order you flagged — re-deriving its deletions against the converted call sites. Resolved in merge 7eae7978: kept #1138's `WithStatement`/`Statement:` conversions everywhere, dropped every `WithBuffered`/`Unbuffered:` scenario, and deleted the converted `TestResumeFromCheckpointPhantom` outright. Verified semantically rather than textually: a grep sweep for the removed APIs (`WithTable(`, `WithAlter(`, `WithBuffered(`, `Unbuffered:`, `copier.Unbuffered`) comes back clean; `go build ./...` and `go vet -tags 'semisync singleversion' ./...` pass (the tagged files are where a stale signature would hide from plain vet); and the tests at every conflict site pass locally — all 20 cutover/rename-column/generated-columns/charset/binary-checksum tests, `TestCheckpoint` under `-tags singleversion`, and the full `pkg/copier` suite (goleak clean).

Auto-merge (squash) is enabled; this lands when checks go green.
